core/vdbe/mod.rs

```diff
@@ -747,97 +308,15 @@ impl Program {
-                                    todo!("{:?}", agg_rhs);
-                                }
-                            },
-                            OwnedValue::Float(lh_f) => match agg_rhs.final_value() {
-                                OwnedValue::Null => OwnedValue::Null,
-                                OwnedValue::Float(rh_f) => {
-                                    OwnedValue::Float(((*lh_f as i64) % (*rh_f as i64)) as f64)
-                                }
-                                OwnedValue::Integer(rh_i) => {
-                                    OwnedValue::Float(((*lh_f as i64) % rh_i) as f64)
-                                }
-                                _ => {
-                                    todo!("{:?}", agg_rhs);
-                                }
-                            },
-                            _ => todo!("{:?}", rhs),
-                        },
-                        (OwnedValue::Agg(aggctx), rhs) => match rhs {
-                            OwnedValue::Integer(rh_i) => match aggctx.final_value() {
-                                OwnedValue::Null => OwnedValue::Null,
-                                OwnedValue::Float(lh_f) => {
-                                    OwnedValue::Float(((*lh_f as i64) % rh_i) as f64)
-                                }
-                                OwnedValue::Integer(lh_i) => OwnedValue::Integer(lh_i % rh_i),
-                                _ => {
-                                    todo!("{:?}", aggctx);
-                                }
-                            },
-                            OwnedValue::Float(rh_f) => match aggctx.final_value() {
-                                OwnedValue::Null => OwnedValue::Null,
-                                OwnedValue::Float(lh_f) => {
-                                    OwnedValue::Float(((*lh_f as i64) % (*rh_f as i64)) as f64)
-                                }
-                                OwnedValue::Integer(lh_i) => {
-                                    OwnedValue::Float((lh_i % (*rh_f as i64)) as f64)
-                                }
-                                _ => {
-                                    todo!("{:?}", aggctx);
-                                }
-                            },
-                            _ => todo!("{:?}", rhs),
-                        },
-                        _ => todo!("{:?} {:?}", state.registers[lhs], state.registers[rhs]),
-                    };
+                    let mut lhs = &state.registers[lhs];
+                    let mut rhs = &state.registers[rhs];
+                    if let OwnedValue::Agg(aggctx) = &lhs {
+                        lhs = aggctx.final_value();
+                    }
+                    if let OwnedValue::Agg(aggctx) = &rhs {
+                        rhs = aggctx.final_value();
+                    }
+                    state.registers[dest] = insn.math_binary_eval()(lhs, rhs);
```

Comment:
Imho I think math function on `OwnedValue` should be outside Insn struct. I don't have a big argument aside from it feels better because insn could be anything else from a math instruction and because we could reuse in other cases. Furthermore, why do we have to parse what type of Math function we do two times, here we know it will be a BitNot, and the we ask math_binary_eval to parse `self` again which in my opinion is unneeded extra work.

Reply:
Yeah I hear you there for sure. Just pushed a revision that has some of that in mind, check it out and lmk what you think.

Reply:
It looks much better now but I think I didn't explain correctly one thing. When you have for example a `BitNot` instruction, you know for sure it will be a `BitNot` operation, but `math_binary_eval` checks a second time what type of instruction it is executing. I preferred a function for each operator so it would be `math_bitnot`, `math_add` , `math..`...

Reply:
Looks like `BitNot` is a good example because that one wasn't extracted but I think it makes the point for other insns